active/003-seq-feature.md

```diff
@@ -0,0 +1,98 @@
+```python
+gene = Feature(name='sagA', type='CDS', strand='+', function='toxin')
+```
+The above code would encode for a gene, whose name is 'sagA', that is a coding region on the positive strand whose function is a toxin.
+
+## `IntervalMetadata` functions
+`constructor(features=None)`
+- `features` : dictionary.  keys are `skbio.Feature` objects.  values are a list of tuples, where each tuple corresponds to an interval.
+- Called to initialize the `IntervalMetadata` object
+
+`reverse_complement(length)`
+- `length` : int.  The length of the genome.  This is required when swapping coordinates
+- This performs a reverse complement on all the coordinates with in IntervalTree.
+
+```python
+   interval_metadata = IntervalMetadata()
+   interval_metadata.add(Feature(gene='sagB', location=0), (3, 5))
+   iv = interval_metadata.reverse_complement(length=10)
+```
+
+`update(old_feature, new_feature)`
+- `old_feature` : `skbio.Feature`.  feature to be replaced
+- `new_feature` : `skbio.Feature`.  updated feature to replace the original feature
+- Since `skbio.Feature` objects are immutable, if individual features in the `IntervalMetadata` are to be updated, the need to be completely replaced.
+
+```python
+    interval_metadata = IntervalMetadata()
+        interval_metadata.add(Feature(gene='sagA', location=0), 1, (4, 7))
+        interval_metadata.add(Feature(gene='sagB', location=0), (3, 5))
+        interval_metadata.update(Feature(gene='sagB', location=0),
+                                 Feature(gene='sagB', location=1))
+```
+
+
+`add(feature, *intervals)`
+- `feature` : `skbio.Feature`. new feature object add into the `IntervalMetadata` object
+- `*intervals` : a list of tuples corresponding to intervals where the feature object exists.
+- This allows for a single feature (include those that have non-continugous features) to be added into the `IntervalMetadata` object.
+
+```python
+   interval_metadata = IntervalMetadata()
+   interval_metadata.add(Feature(gene='sagA', location=0), 1, (4, 7))
+   interval_metadata.add(Feature(gene='sagB', location=0), (3, 5))
+```
+
+`query(*args, **kwargs)`
+- `*args` : a list of tuples to search for features
+- `**kwargs` : keyword arguments to query features by their attributes.
+- This allows for features to be searched by both intervals and keywords (i.e. CDS vs exon, ...)
+
+```python
+   interval_metadata = IntervalMetadata(features={
+                                        Feature(gene='sagA', location=0): [(0, 2), (4, 7)],
+                                        Feature(gene='sagB', location=0): [(3, 5)]})
+ 
+   feats = interval_metadata.query((1, 2))
+   feats = interval_metadata.query(gene='sagB')
+```
+
+This is the current implementation in micronota and fits all of our use-cases at the moment.  We don't feel that IntervalIndex objects as suggested in the recent pandas poll request can easily handle all of our use cases (i.e. non-contiguous intervals isn't straightforward).
```

Comment:
Forgot to add this in the RFC.

We are actually handling that scenario - if you have the features, you can
lookup the associated intervals, and slice the Sequence Object by those
intervals.  In the original PR, the __getitem__ method was modified to
accept Features as input.

On Thu, Apr 14, 2016 at 9:28 AM, Evan Bolyen <notifications@github.com>
wrote:

> In active/003-seq-feature.md
> <https://github.com/biocore/scikit-bio-rfcs/pull/9#discussion_r59748063>:
>
> > +
> > +`query(*args, **kwargs)`
> > +- `*args` : a list of tuples to search for features
> > +- `**kwargs` : keyword arguments to query features by their attributes.
> > +- This allows for features to be searched by both intervals and keywords (i.e. CDS vs exon, ...)
> > +
> > +```python
> > +   interval_metadata = IntervalMetadata(features={
> > +                                        Feature(gene='sagA', location=0): [(0, 2), (4, 7)],
> > +                                        Feature(gene='sagB', location=0): [(3, 5)]})
> > +
> > +   feats = interval_metadata.query((1, 2))
> > +   feats = interval_metadata.query(gene='sagB')
> > +```
> > +
> > +This is the current implementation in micronota and fits all of our use-cases at the moment.  We don't feel that IntervalIndex objects as suggested in the recent pandas poll request can easily handle all of our use cases (i.e. non-contiguous intervals isn't straightforward).
>
> It would be nice if there was a way to slice the sequence by one or more
> features. Right now the feature is pretty decoupled from it's location, so
> I don't see a straight-forward way to accomplish this.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/biocore/scikit-bio-rfcs/pull/9/files/b954b92abe3eb629826cd2a898461e97e57dd057#r59748063>
>